### PR TITLE
storage: lock raftMu in Replica.proposeRaftCommand

### PR DESCRIPTION
**Pull request by petermattis**
See #9665.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9814)

### CODE REVIEW

**Comment by tschottdorf** [comment]
LGTM but #9655 isn't the correct issue (which is good because I was going to suggest to leave the issue open to find good ways to avoid regression).

File: storage/replica.go [hunk 1]
```diff
@@ -1670,6 +1674,17 @@ func makeIDKey() storagebase.CmdIDKey {
...
 ) (chan roachpb.ResponseWithError, func() bool, error) {
+	// proposePendingCmdLocked calls withRaftGroupLocked which requires that
+	// raftMu is held. In order to maintain our lock ordering we need to lock
+	// Replica.raftMu here before locking Replica.mu.
+	//
+	// TODO(peter): It appears that we only need to hold Replica.raftMu when
+	// calling raft.NewRawNode. We could get fancier with the locking here to
+	// optimize for the common case where Replica.mu.internalRaftGroup is
+	// non-nil, but that doesn't currently seem worth it. Always locking raftMu
+	// has a tiny (~1%) performance hit.
```
Comment (petermattis) [inline storage/replica.go]: Done.

**Comment by petermattis** [comment]
Fixed the commit message to point to #9665 and leave it open.

**Comment by tamird** [comment]
Don't forget to also update the PR description or GH will close the issue anyway.

**Comment by tschottdorf** [comment]
Looks all set. LGTM.

**Comment by tamird** [comment]
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 1196 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814#-KTWZ0Gxnpn3jxU70WC-:-KTWZ0Gxnpn3jxU70WC0:b62j9dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd69e1550791761595810ba68d0c5d84a6a55d95/storage/client_raft_test.go#L1196)):*
> ```Go
> func runReplicateRestartAfterTruncation(t *testing.T, removeBeforeTruncateAndReAdd bool) {
> 	sc := storage.TestStoreConfig()
> 	// Don't timeout raft leaders or range leases. This test expects
> ```

How does this prevent timing out range leases? Maybe I'm misunderstanding the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814)*

**Comment by petermattis** [comment]
@bdarnell PTAL at this when you return and let me know if you have any better thoughts as to how to fix it.

File: storage/replica.go [hunk 2]
```diff
@@ -382,6 +382,8 @@ var _ KeyRange = &Replica{}
 // handleRaftReady for an instance of where this value varies. The
 // shouldCampaign argument indicates whether a new raft group should be
 // campaigned upon creation and is used to eagerly campaign idle replicas.
+//
+// Requires that Replica.raftMu is held.
```
Comment (bdarnell) [inline storage/replica.go]: Specify that this method requires both `Replica.raftMu` and `Replica.mu`.

**Review by bdarnell — APPROVED** [review]

File: storage/replica.go (hunk 2 above)
Comment (petermattis) [inline storage/replica.go]: Done.

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 1196 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814#-KTWZ0Gxnpn3jxU70WC-:-KTdvAZ3gcs6b8NBtBdk:bu47789) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd69e1550791761595810ba68d0c5d84a6a55d95/storage/client_raft_test.go#L1196)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> How does this prevent timing out range leases? Maybe I'm misunderstanding the comment.
</details>
The lease duration is 3x the raft election timeout. The raft election timeout is `RaftElectionTimeoutTicks * RaftTickInterval`. So making `RaftElectionTimeoutTicks` really large makes the lease duration really large.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 1196 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814#-KTWZ0Gxnpn3jxU70WC-:-KTe4yoW2RM51OIWQT2M:b36zcdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd69e1550791761595810ba68d0c5d84a6a55d95/storage/client_raft_test.go#L1196)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The lease duration is 3x the raft election timeout. The raft election timeout is `RaftElectionTimeoutTicks * RaftTickInterval`. So making `RaftElectionTimeoutTicks` really large makes the lease duration really large.
</details>
I see. I think it'd be helpful to spell out this relationship in this comment since it may change in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9814)*

**Pull request merged**